### DISCUSSION TITLE
How can I get the sum of the rows on all the pages and show them?

### DISCUSSION THREAD

**Discussion by SebnemSeymen** [serenity-is/Serenity#6792] [Help]
I am using pagination in my list. Because I have too many rows. I can show the sum of the columns in the footer of my list. But since there is pagination, only the sum of the relevant list comes up. For example, if there is a yield of 2 pages, if the total on the first page is 500, and if the total on the second page is 100, I want to show the grand total as 600.

The code I use is as follows;

       this.view.setSummaryOptions({
                 aggregators: [

                     new Slick.Aggregators.Sum(fld.AMOUNT),
                     new Slick.Aggregators.Sum(fld.VAT),
                     new Slick.Aggregators.Sum(fld.TOTALAMOUNT)
                 ]
             });

![Ekran Görüntüsü (151)](https://github.com/serenity-is/Serenity/assets/121672307/cbfbe2fa-8b30-4b00-993b-62fc6f573344)

Alternatively, I also use:

    (this.view as any).onDataChanged.subscribe(() => {
                 var total = 0;
                 for (var k of this.view.getItems()) {
                     total += k.AMOUNT || 0;
                 }

                 console.log(total)
                 Q.notifySuccess(total.toString());
             });

I will be glad if you help.

![Ekran Görüntüsü (152)](https://github.com/serenity-is/Serenity/assets/121672307/232715c1-fa45-4762-b068-7e2cb4097ac8)

**Comment by wezmag** [comment]
Hello @SebnemSeymen 
It is a very interesting topic. 
When using `Aggregators.Sum`, the calculation happens on the client side. 
That's the reason why the sum does not add up across pages with a paged query.
To overcome the limitation, you need to get the total on the server side, then display the result in the footer column of the grid.

I will use the Product table in the Northwind database as an example.
### Get total on the server side.
Find `ProductsListHandler.cs` file in the RequestHandlers folder, and override the `OnReturn` method as below
``` csharp
    protected override void OnReturn()
    {
        base.OnReturn();
        // Get total units in stock and return it with the Response object
        Response.CustomData = new()
        {
            { "TotalUnitsInStock", GetTotalUnitsInStock() }
        };
    }

    private decimal GetTotalUnitsInStock()
    {
        var query = new SqlQuery()
            .Select(Sql.Sum(MyRow.Fields.UnitsInStock))
            .From(MyRow.Fields.TableName, new Alias("T0"));
        // In order to get the correct result, you want to ensure the same filters are applied to this query by calling ApplyFilters
        ApplyFilters(query);
        // change _logger to null, if you did not inject ILogger here.
        var total = SqlHelper.ExecuteScalar(Connection, query, _logger); 
        return total == DBNull.Value ? 0 : Convert.ToDecimal(total);
    }
```

### Display the result on the footer row of the grid.
Find `ProductsGrid.ts` file in the Products folder, and add code below
``` typescript
private total: number = 0;

constructor(container: JQuery) {
    super(container);

    // When data loaded, display total in the footer column.
    this.view.onDataLoaded.subscribe((e, args) => {
        const columnIndex = indexOf(this.getColumns(), c => c.field === ProductsRow.Fields.UnitsInStock);
        this.slickGrid.getFooterRowColumn(columnIndex).textContent = "Total:" + this.total.toString();
    });
}

protected onViewProcessData(response) {
    const res = super.onViewProcessData(response);
    // Get the TotalUnitsInStock from the response
    this.total = (res as any).CustomData["TotalUnitsInStock"];
    return res;
}

protected getSlickOptions() {
    const opt = super.getSlickOptions();
    // Display footer row on the grid.
    opt.showFooterRow = true;
    return opt;
}
```

Serene version @6.7.1

**Reply by SebnemSeymen** [reply]
thank your response ;i am trying your solve

           private decimal GetTotalUnitsInStock()
            {
                var queryTotal = new SqlQuery()
                 .From(MyRow.Fields).Select(Sql.Sum(MyRow.Fields.AMOUNT.Expression)
                 );
                ApplyFilters(queryTotal);

                var total = SqlHelper.ExecuteScalar(Connection, queryTotal);


                return total == DBNull.Value ? 0 : Convert.ToDecimal(total);
            }

this function returns the sum without any problem.

        Response.CustomData = new()
        {
            { "TotalUnitsInStock", GetTotalUnitsInStock() }
        };



This part gives an error. I cannot add CustomData next to Entities, TotalCouunt, Take, Skip parameters. Since I cannot add it to ListResponse, it cannot find it in Grid.ts either. Is there any other way to send this value to Grid.ts?

**Reply by wezmag** [reply]
It seems like you are using an older version that doesn't have `CustomData` property.
You can add another ListResponse that extends the original ListResponse.
For example:
```  csharp
public class YourNewListResponse : ListResponse<MyRow>
{
    public decimal TotalUnitsInStock { get; set; }
}
```
Then in the ListHandler, change `ListResponse<MyRow>` to `YourNewListResponse`.
You should be able to add your data to the new property.
``` csharp
protected override void OnReturn()
{
    base.OnReturn();
    //Get total units in stock and return it with the Response object
    //Response.CustomData = new()
    //{
    //    { "TotalUnitsInStock", GetTotalUnitsInStock() }
    //};
   Response.TotalUnitsInStock = GetTotalUnitsInStock();
}
```
On the client side:
``` ts
protected onViewProcessData(response) {
    const res = super.onViewProcessData(response);
    //this.total = (res as any).CustomData["TotalUnitsInStock"];
    this.total = (res as YourNewListResponse).TotalUnitsInStock;
    return res;
}
```

**Reply by SebnemSeymen** [reply]
public class YourNewListResponse : ListResponse<MyRow>
{
    public decimal TotalUnitsInStock { get; set; }
}

Should I add this in repository.cs?

**Reply by wezmag** [reply]
`Repository.cs`? That's very old version.
I believe you can put the code in the `Respository.cs` file.
However, do not put the code in side the Repository class.
Don't forget to run T4 templates, after adding the new ListResponse class.

**Reply by SebnemSeymen** [reply]
yes, a very old version was installed. I just took over the project. so i am very new to serenity. Could you please explain what T4 is and how to run it?

**Reply by wezmag** [reply]
T4 templates are for code generation. As Victor said, you are using .net 3.x version, you don't need to run T4.

**Reply by wezmag** [reply]
For your version, you need to modify the `repository.cs` as below
``` csharp
//...
//public ListResponse<MyRow> List(IDbConnection connection, ListRequest request)
public YourNewListResponse List(IDbConnection connection, ListRequest request)
{
    return new MyListHandler().Process(connection, request);
}
//...
//private class MyListHandler : ListRequestHandler<MyRow>
private class MyListHandler : ListRequestHandler<MyRow, ListRequest, YourNewListResponse> {
    protected override void OnReturn()
    {
        base.OnReturn();
        Response.TotalUnitsInStock = GetTotalUnitsInStock();
    }
}
//...
```

**Reply by SebnemSeymen** [reply]
this worked for me. Thank you for your patience and interest. I'm closing the issue. have a nice day.

**Reply by wezmag** [reply]
If my answer solved your problem, please mark it. It will benefit others who have the same question as you.

**Comment by SebnemSeymen** [comment]
![Ekran Görüntüsü (157)](https://github.com/serenity-is/Serenity/assets/121672307/ce9386f1-2fe7-44ff-9bee-be7097bfb883)

i m using vs 2022

**Reply by SebnemSeymen** [reply]
"Build" => "Transform All T4 Templates dont show at VS 2022

**Reply by VictorTomaili** [reply]
You are using .net 3.x version, which is uses sergen instead of t4 templates. Transform is completed successfully, as I see in your logs. Check `Transforming` text in your build log.

**Reply by SebnemSeymen** [reply]
if my tables were transformed then i should be able to use Right now it won't let me define it.
   `     private class MyListHandler : ListRequestHandler<MyRow> 
        {
                protected override void OnReturn()
                {
                    base.OnReturn();
                    Response.GrandTotal= GetTotalUnitsInStock();
                }
        }
`

![Ekran Görüntüsü (158)](https://github.com/serenity-is/Serenity/assets/121672307/d88f3095-d2a7-4d0f-9e64-1065e18ea5c7)
